core/vm/evm.go

```diff
@@ -174,6 +174,21 @@ func (evm *EVM) SetBlockContext(blockCtx BlockContext) {
 // the necessary steps to create accounts and reverses the state in case of an
 // execution error or failed value transfer.
 func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error) {
+	if tracer := evm.Config.Tracer; tracer != nil {
+		if evm.depth == 0 {
+			tracer.CaptureStart(caller.Address(), addr, false, input, gas, value)
+			defer func(startGas uint64) { // Lazy evaluation of the parameters
+				tracer.CaptureEnd(ret, startGas-leftOverGas, err)
+			}(gas)
+		} else {
+			// Handle tracer events for entering and exiting a call frame
+			tracer.CaptureEnter(CALL, caller.Address(), addr, input, gas, value)
+			defer func(startGas uint64) {
+				tracer.CaptureExit(ret, startGas-leftOverGas, err)
+			}(gas)
+		}
```

Comment:
I don't personally see why those are not the same call with and extra parameter to determine the `depth` (so root is `depth == 0`).

There is lot of duplication of in those code that could be merged into a single statement. I think it's reasonable if this change is applied (e.g. move the capture higher) than it's a good time to change the interface. There is higher chance that current user's of tracer will read a bit to understand what to change and they will probably catch at the same time that the call is now made higher.